Sources/GRPCCore/Call/Client/CallOptions.swift

```diff
@@ -80,33 +80,33 @@ public struct CallOptions: Sendable {
   /// Whether compression is enabled or not for request and response messages.
   public var compression: Compression
 
+  /// Compression configuration.
+  ///
+  /// Note that this configuration is advisory: not all transports support compression and may
+  /// ignore this configuration. Transports which support compression will use this configuration
+  /// in preference to any compression configured at a transport level.
   public struct Compression: Sendable {
-    /// Whether request messages should be compressed.
+    /// The algorithm used for compressing outbound messages.
     ///
-    /// Note that this option is _advisory_: transports are not required to support compression.
-    public var requests: Bool
+    /// If `nil` the value configured on the transport will be used instead.
+    public var algorithm: CompressionAlgorithm?
 
-    /// Whether response messages are permitted to be compressed.
-    public var responses: Bool
+    /// The enabled compression algorithms.
+    ///
+    /// If `nil` the value configured on the transport will be used instead.
+    public var enabledAlgorithms: CompressionAlgorithmSet?
```

Comment:
That's fair. My aim was to make this API simpler but don't think I achieved that with the current semantics.

I looked into other implementations some more and they _don't_ expose the enabled set at the call level:

- Java: [`withCompression`](https://grpc.github.io/grpc-java/javadoc/io/grpc/CallOptions.html#withCompression(java.lang.String))
- Python: [`compression=...`](https://github.com/grpc/grpc/tree/master/examples/python/compression)
- C++: [`set_compression_algorithm`](https://github.com/grpc/grpc/blob/84ee28e6956ed7cd51462aad52e64782dd5ca34b/examples/cpp/compression/greeter_client.cc#L59)

They also use the set of algorithms to validate that the selected algorithm for requests is allowed, i.e. if you set the enabled algorithms to be "gzip" and "none" on the transport, and select "deflate" at the call level, then the request won't be compressed, which makes sense.

So I think what I'm proposing here is to reframe "enabled response algorithms" from "algorithms we support for decompressing responses" to "algorithms we support". More concretely:
- remove `enabledAlgorithms` from the call level, and only allow it to be set at transport creation time
- keep `algorithm` and note that it may not be applied if it isn't enabled on the transport
- have a default compression algorithm config on the transport

This also puts it inline with other implementations. WDYT?

Reply:
Sounds good, thanks for doing the research.